examples/snake/snake.cc

```diff
@@ -0,0 +1,651 @@
+// Copyright lowRISC Contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cheri.hh>
+#include <compartment.h>
+#include <debug.hh>
+#include <platform-gpio.hh>
+#include <thread.h>
+#include <vector>
+
+#include "../../libraries/lcd.hh"
+#include "cherry_bitmap.h"
+#include "xor_shift.hh"
+
+using Debug = ConditionalDebug<true, "Snake">;
+using namespace sonata::lcd;
+using namespace CHERI;
+
+// Control game speed
+static constexpr uint32_t MILLISECONDS_PER_FRAME = 400;
+// Small wait between games to avoid accidentally starting the next
+static constexpr uint32_t START_MENU_WAIT_MILLISECONDS = 400;
+
+// Snake speeds up as it gets longer if enabled
+static constexpr bool SPEED_SCALING_ENABLED = true;
+// If enabled, all joystick motions start the game (not just a press)
+static constexpr bool START_ON_ANY_INPUT = true;
+// If enabled, displays a cherry bitmap for the fruit instead of a
+// green square at 10x10 and 5x5 tile sizes.
+static constexpr bool USE_CHERRY_IMAGE = true;
+
+// Change colour of game elements
+static constexpr Color BACKGROUND_COLOR = Color::Black,
+                       BORDER_COLOR     = Color::White,
+                       FOREGROUND_COLOR = Color::White,
+                       SNAKE_COLOR      = Color::Red;
+// Change the size of game elements (will automatically fit display)
+static constexpr Size TILE_SIZE    = {10, 10}, 
+                      TILE_SPACING = { 2,  2},
+                      BORDER_SIZE  = { 4,  3};
+
+typedef struct Position
+{
+	int32_t x;
+	int32_t y;
+} Position;
+
+enum class Direction
+{
+	UP    = 0,
+	RIGHT = 1,
+	DOWN  = 2,
+	LEFT  = 3
+};
+
+enum class Tile : uint64_t
```

Comment:
Why uint64_t?

Reply:
We looked into why this was needed - the memory allocater mallocs in chunks of 8 bytes and so certain sizes could cause some issues with the capabilities detecting out of bounds accesses. For now for the purposes of this demo I've just left this as uint64_t to make sure everything works.